### DISCUSSION TITLE
wall ar-placement for ios

### DISCUSSION THREAD

**Discussion by PikabobAlex** [google/model-viewer#2466] [Q&A]
Hi community, is ios not supporting wall placement feature yet? I've enabled it in my model-viewer attribute but seems like only Andoird has it but not ios

**Comment by ThorstenBux** [comment]
Works fine for me. However, you need to create your USDZ manually and set it to wall placement in Reality Composer. The automatically created USDZ doesn't seem to be able to do it.

**Reply by PikabobAlex** [reply]
but what if I want to set custom texture in html? Texture are not premade

**Reply by PikabobAlex** [reply]
yeah, just tested it out, it works, but I unable to change the texture on the fly. Thanks tho.

**Comment by ThorstenBux** [comment]
Yes when you change the texture I assume you do that change to the GLB file and then pass it to ModelViewer. And when ModelViewer converts the file to USDZ it looses the information about the wall tracking. Which brings me to the question I asked here: https://discourse.threejs.org/t/usdz-exporter-vertical-tracking/26991 how to encode inside the USDZ that we want wall tracking? Is it upAxis set to “z”? Don't know where to read up on the USDZ file format

**Reply by elalish** [reply]
Yes, this is a Safari bug as far as I'm aware. They don't support query params on Blob URLs, so their own Quick Look API is defeated.

**Reply by felipe-chamas** [reply]
From what I've tested, quicklook does not care about vertical placement properties on the USDZ model and snaps it to any plane found (For devices >= iPhone X). So you just need to move the model to the wall in order to place it. Older devices seem to work only with floor placement and maybe wall placement with ARKit (but not quicklook).
Once again, their documentation is just terrible to confirm this... But I would like very much to be wrong

**Comment by milesgreen** [comment]
My experience is that wall placement does work on iOS. But with slight limitations....

I have tested as far back as iPhone 7 (on iOS 14)
I have tried it with ModelViewer loading a GLB with Material Variants and dynamically generating the USDZ.
I have also tested it with ModelViewer loading in a GLB, then loading an external texture map, and dynamically generating a USDZ.
Both work. I am able to place the generated USDZs on a wall with iOS QuickLook.

The issue is, I don't think AR QuickLook has a property to specify horizontal / vertical placement. At least, not when invoking QuickLook from web. QuickLook will simply try to anchor it to whatever surface you scan. Regardless of what ModelViewer AR placement you go for (floor or wall) QuickLook will allow both to anchor to either. So there is no way to specify _just_ floor placement or _just_ wall placement for QuickLook.

Also, as with any AR, vertical placement is always a bit trickier due to the lack of surface detail. It can sometimes look as though it's failing if you're trying to place an object on a completely blank wall. Later devices seem to handle this progressively better, but the iPhone 12 Pro with its LiDAR Scanner is spot on.

**Comment by ThorstenBux** [comment]
Made the changes and opened a PR on THREE.js for the USDZExport. https://github.com/mrdoob/three.js/pull/21997
